chore: calling other djangoapps from API instead of model

deborahgu opened the pull request:
## Description

i converted the  Programs rest  API views from calling other djangoapps via direct model access to calling from API. This included  adding an API in the Student app.

Should be invisible to all  learners. Other developers will now have access to the new API method.

FIXES: APER-3972

deborahgu committed:
chore: calling other djangoapps from API instead of model

switching  from calling other djangoapps via direct model access to
calling from API. This included  adding an API in the Student app.

FIXES: APER-3972

deborahgu commented on common/djangoapps/student/api.py:
```diff
@@ -92,13 +97,7 @@ def create_manual_enrollment_audit(
     else:
         enrollment = None
 
-    _create_manual_enrollment_audit(
```
I just let my IDE reformat  according to linting rules; since it was only a couple of minor changes I didn't prevent it from doing that.

deborahgu commented on common/djangoapps/student/api.py:
```diff
@@ -132,7 +131,31 @@ def is_user_staff_or_instructor_in_course(user, course_key):
-        GlobalStaff().has_user(user) or
-        CourseStaffRole(course_key).has_user(user) or
-        CourseInstructorRole(course_key).has_user(user)
+        GlobalStaff().has_user(user)
```
autolint again

deborahgu commented on common/djangoapps/student/tests/test_api.py:
```diff
@@ -33,10 +39,7 @@ def test_is_user_enrolled_in_course(self):
-            user_id=self.user.id,
-            course_id=self.course.id
-        )
+        CourseEnrollmentFactory.create(user_id=self.user.id, course_id=self.course.id)
```
autolint again

deborahgu commented on openedx/core/djangoapps/programs/rest_api/v1/views.py:
```diff
@@ -170,26 +173,22 @@ def transform_authoring_organizations(authoring_organizations) -> list[dict[str,
+        course_enrollments = get_course_enrollments(user, True, (enterprise_enrollment_course_ids))
 
-        return list(course_enrollments)
+        return course_enrollments
```
I'm not a big fan of passing around lists, when you could pass around a QuerySet, so I converted this to return the QuerySet. If I had my druthers the `ProgramProgressMeter` would also take a QuerySet, but honestly at some point that whole thing is going to get re-factored in some way or another.

justinhynes commented on openedx/core/djangoapps/programs/rest_api/v1/views.py:
```diff
@@ -170,26 +173,22 @@ def transform_authoring_organizations(authoring_organizations) -> list[dict[str,
-            .filter(course_id__in=enterprise_enrollment_course_ids)
-            .select_related("course")
-        )
+        course_enrollments = get_course_enrollments(user, True, (enterprise_enrollment_course_ids))
```
Quick question -- It looks like the new `get_course_enrollments` function is expecting `course_ids` to be a list (type-hint: `course_ids: list[str | None] | None = None,`), but we no longer are casting `enterprise_enrollment_course_ids` to a list on line 183 in the updated `_get_enterprise_course_enrollments` function. 



It looks like `enterprise_enrollment_course_ids` will now be a QuerySet. Is this okay? I think it's fine? since the `course_enrollments = course_enrollments.filter(course_id__in=course_ids)` line just needs an iterable? I don't feel strongly either way in terms of fixing the type hint or explicitly casting `enterprise_enrollment_course_ids` as a list. 

Let me know what you think!

deborahgu replied:
I think you're right, and it should be a List. In fact, I think that's probably my original intention, which is why there is a meaningless set of parens  around the passed parameter  – I think I  started to cast it and then left off the word `list`.

 I thought about sending it as a QuerySet, which I prefer, but the reality is I don't want the caller to have an idea about what kind of course ID list might be getting sent over. In theory, it could be someone just sending over a list of a couple of course IDs to search through.

 ( python's loose typing is very forgiving of treating QuerySets as  list switches I assume why this worked, but it really shouldn't be. If I had enabled the type checker on `student/api.py` i'm sure it would have yelled at me about that.)

deborahgu committed:
chore: casting a query set

Casting the query set to list, because the caller is not making any
assumptions about what kind of IDs are being sent.

deborahgu committed:
Merge branch 'master' into dkaplan1/APER-3972_program-dashboard-api-remove-direct-links-to-models
